Mathlib/Tactic/TermCongr.lean

```diff
@@ -0,0 +1,695 @@
+        }but its right-hand side is not definitionally equal to the expected value{indentD rhs}"
+    return pf
+
+/-- Force the lhs and rhs to be defeq. For when `dsimp`-like congruence is necessary.
+Clears the proof. -/
+def CongrResult.defeq (res : CongrResult) : MetaM CongrResult := do
+  if res.isRfl then
+    return res
+  else
+    unless ← isDefEq res.lhs res.rhs do
+      throwError "Cannot generate congruence because we need{indentD res.lhs}\n{""
+        }to be definitionally equal to{indentD res.rhs}"
+    -- Propagate types into any proofs that we're dropping:
+    discard <| res.eq
+    return {res with pf? := none}
+
+/-- Tries to make a congruence between `lhs` and `rhs` automatically.
+1. If they are defeq, returns a trivial congruence.
+2. Tries using `Subsingleton.elim`.
+3. Tries `proof_irrel_heq` as another effort to avoid doing congruence on proofs.
+3. Otherwise throws an error.
+
+Note: `mkAppM` uses `withNewMCtxDepth`, which prevents `Prop` specialization.
+Otherwise the `Subsingleton Prop` instance could apply for example. -/
+def CongrResult.mkDefault (lhs rhs : Expr) : MetaM CongrResult := do
+  if ← isDefEq lhs rhs then
+    return {lhs, rhs, pf? := none}
+  else if let some pf ← (observing? <| mkAppM ``Subsingleton.elim #[lhs, rhs]) then
+    return CongrResult.mk' lhs rhs pf
+  else if let some pf ← (observing? <| mkAppM ``proof_irrel_heq #[lhs, rhs]) then
+    return CongrResult.mk' lhs rhs pf
+  throwError "Could not generated congruence between{indentD lhs}\nand{indentD rhs}"
+
+/-- Does `CongrResult.mkDefault` but makes sure there are no lingering congruence holes. -/
+def CongrResult.mkDefault' (mvarCounterSaved : Nat) (lhs rhs : Expr) : MetaM CongrResult := do
+  if let some h := hasCHole mvarCounterSaved lhs then
+    throwError "Left-hand side{indentD lhs}\nstill has a congruence hole{indentD h}"
+  if let some h := hasCHole mvarCounterSaved rhs then
+    throwError "Right-hand side{indentD rhs}\nstill has a congruence hole{indentD h}"
+  CongrResult.mkDefault lhs rhs
+
+/-- Throw an internal error. -/
+def throwCongrEx (lhs rhs : Expr) (msg : MessageData) : MetaM α := do
+  throwError "congr(...) failed with left-hand side{indentD lhs}\n{""
+    }and right-hand side {indentD rhs}\n{msg}"
+
+/-- If `lhs` or `rhs` is a congruence hole, then process it.
+Only process ones that are at least as new as `mvarCounterSaved`
+since nothing prevents congruence holes from leaking into the local context. -/
+def mkCongrOfCHole? (mvarCounterSaved : Nat) (lhs rhs : Expr) : MetaM (Option CongrResult) := do
+  match cHole? lhs mvarCounterSaved, cHole? rhs mvarCounterSaved with
+  | some (isLhs1, val1, pf1), some (isLhs2, val2, pf2) =>
+    trace[Elab.congr] "mkCongrOfCHole, both holes"
+    unless isLhs1 == true do
+      throwCongrEx lhs rhs "A RHS congruence hole leaked into the LHS"
+    unless isLhs2 == false do
+      throwCongrEx lhs rhs "A LHS congruence hole leaked into the RHS"
```

Comment:
Do these actually happen? I'm unclear if these are meant to be considered unreachable or just unfortunate. :-)

Reply:
If it happens, it's really unfortunate, but I wouldn't expect it in normal operation. The expected type could be an equality with a metavariable shared between the LHS and RHS, letting a hole leak from the LHS to the RHS.

I don't really know how else this can occur in practice, but I figure one day someone on Zulip will let us know :wink: